idx/memory/memory.go

```diff
@@ -484,49 +534,46 @@ func (m *UnpartitionedMemoryIdx) indexTags(def *schema.MetricDefinition) {
 // a return value of "false" means there was an error and the deindexing was
 // unsuccessful, "true" means the indexing was at least partially or completely
 // successful
-func (m *UnpartitionedMemoryIdx) deindexTags(tags TagIndex, def *schema.MetricDefinition) bool {
-	for _, tag := range def.Tags {
-		tagSplits := strings.SplitN(tag, "=", 2)
-		if len(tagSplits) < 2 {
-			// should never happen because every tag in the index
-			// must have a valid format
-			invalidTag.Inc()
-			log.Errorf("memory-idx: Tag %q of id %q has an invalid format", tag, def.Id)
-			continue
-		}
-
-		tagName := tagSplits[0]
-		tagValue := tagSplits[1]
-		tags.delTagId(tagName, tagValue, def.Id)
+func (m *UnpartitionedMemoryIdx) deindexTags(tags TagIndex, def *idx.MetricDefinition) bool {
+	for _, tag := range def.Tags.KeyValues {
+		tags.delTagId(tag.Key, tag.Value, def.Id)
 	}
 
-	tags.delTagId("name", def.Name, def.Id)
+	nameKey, _ := idx.IdxIntern.GetPtrFromByte([]byte("name"))
+	nameValue, _ := idx.IdxIntern.GetPtrFromByte([]byte(def.Name.String()))
+	tags.delTagId(nameKey, nameValue, def.Id)
 
 	m.defByTagSet.del(def)
 
 	return true
 }
 
-// Used to rebuild the index from an existing set of metricDefinitions for a specific paritition.
-func (m *UnpartitionedMemoryIdx) LoadPartition(partition int32, defs []schema.MetricDefinition) int {
+// LoadPartition is used to rebuild the index from an existing set of metricDefinitions for a specific paritition.
+func (m *UnpartitionedMemoryIdx) LoadPartition(partition int32, defs []idx.MetricDefinition) int {
 	// UnpartitionedMemoryIdx isnt partitioned, so just ignore the partition passed and call Load()
 	return m.Load(defs)
 }
 
-// Used to rebuild the index from an existing set of metricDefinitions.
-func (m *UnpartitionedMemoryIdx) Load(defs []schema.MetricDefinition) int {
+// Load is used to rebuild the index from an existing set of metricDefinitions.
+func (m *UnpartitionedMemoryIdx) Load(defs []idx.MetricDefinition) int {
 	m.Lock()
 	defer m.Unlock()
 	var pre time.Time
 	var num int
 	for i := range defs {
 		def := &defs[i]
+		nDef := *def
 		pre = time.Now()
 		if _, ok := m.defById[def.Id]; ok {
 			continue
 		}
 
-		m.add(def)
+		m.add(&nDef)
+
+		if TagSupport {
+			// create new def to avoid holding open the backing array of defs which is passed up from the persistent index
+			m.indexTags(&nDef)
```

Comment:
didn't `add()`, which has just been called, already index those tags?

Reply:
Yes, I think some things have changed and I might have missed it during one of the rebases